### DISCUSSION TITLE
Why exactly(other than disabled by kernel) doesn't swap work in TerraOS and Shimboot? How could we get around this(like building the image a certain way or modifying the kernel of the chromeos rma shim or something)?

### DISCUSSION THREAD

**Discussion by FunnyCorgi** [3kh0/ext-remover#1194] [Help JOIN TITANIUM NETWORK TO SHARE PLEASE] [answered]
Why exactly(other than disabled by kernel) doesn't swap work in TerraOS and Shimboot? How could we get around this(like building the image a certain way or modifying the kernel of the chromeos rma shim or something)?

**Comment by HelloIloveGimkit2** [comment]
Swap functionality in operating systems like TerraOS and Shimboot can be limited for several reasons, primarily related to the architecture and design choices of the OS or the bootloader. Here are some possible reasons and suggestions for workarounds:

### Reasons for Swap Limitations

1. **Kernel Configuration**: The kernel may not have been configured to support swap. This can be due to missing swap-related options or drivers during the kernel build process.

2. **Filesystem Limitations**: The filesystem used may not support swap files or partitions, or there could be issues with the way swap is defined in the filesystem.

3. **Memory Management**: The memory management subsystem may not be fully implemented or optimized for swap operations, leading to inefficiencies or failures when trying to use swap space.

4. **Boot Environment**: Shimboot may not provide the necessary environment or configurations to access swap. It might not initialize memory management features that are required for swapping.

5. **Hardware Constraints**: The underlying hardware may not support certain types of swapping (e.g., if the hardware doesn’t allow for virtual memory management).

### Workarounds

1. **Kernel Compilation Options**: Ensure that when you build the kernel, you include support for swap. This often involves enabling specific configurations related to memory management and swap support.

2. **Using a Different Filesystem**: If the current filesystem doesn’t support swap, consider using one that does. For example, ext4 generally has good support for swap files.

3. **Configuring Swap Partitions**: Create a dedicated swap partition instead of a swap file, as this can sometimes lead to better performance and reliability.

4. **Boot Parameters**: Check if there are boot parameters in Shimboot that need to be set to enable or properly configure swap functionality.

5. **Testing with Minimal Configurations**: Start with a minimal configuration of the OS that includes basic features. Once swap works in this configuration, gradually add more features to identify what breaks it.

6. **Debugging**: Use debugging tools or logs to check where the swap functionality fails. This can provide insights into what specific components are causing issues.

7. **Consult Documentation**: Review any available documentation for both TerraOS and Shimboot regarding swap functionality to see if there are known limitations or specific instructions for enabling it.

8. **Community Support**: If the OS is part of an open-source project, consider reaching out to the community for support. Others may have encountered and solved similar issues.

By addressing these aspects, you can potentially enable swap functionality in TerraOS and Shimboot more effectively.

generated by chatgpt lmao I'm lazy

**Reply by HelloIloveGimkit2** [reply]
don't think that chatgpt fucking knows what terraos or shimboot is

**Reply by FunnyCorgi** [reply, marked as answer]
> don't think that chatgpt fucking knows what terraos or shimboot is

@HelloIloveGimkit2 I gave it links and it looked at them, then it gave me this:

Swap doesn't work in TerraOS and Shimboot primarily because **swap is disabled in the kernel**. Both of these projects involve exploiting vulnerabilities in Chromebook RMA shims to boot Linux-based operating systems in environments originally designed for recovery and diagnostics. These environments were not intended to support typical Linux functionalities like swap or deep sleep, and thus the kernel used in these systems has swap functionality explicitly disabled. Here’s a more detailed look into why and how you can get around this:

### Why Swap is Disabled:
1. **Minimalistic Design**: TerraOS and Shimboot are designed to run in a lightweight environment, using Chromebook RMA shims to bypass restrictions. These shims are built for recovery purposes, meaning many features typical in full Linux distributions (like swap) are omitted to minimize the complexity of the environment.
2. **Kernel-Level Disablement**: In TerraOS, it's specifically mentioned that swap is disabled at the kernel level. ChromeOS-based environments like these often restrict functionality to improve system stability and security in their recovery environments【12†source】【13†source】.

### Workarounds:
1. **Recompile the Kernel**: The most straightforward way to enable swap is by **modifying the kernel configuration** to re-enable swap support. You would need to:
   - Modify the kernel’s `.config` file and ensure that `CONFIG_SWAP` is set to `y`.
   - Recompile the kernel with these settings and include it in your custom TerraOS or Shimboot build.
   - Rebuild the OS image following the instructions in the [TerraOS GitHub repository](https://github.com/r58Playz/terraos)【13†source】.

2. **Custom OS Image with Swap Support**: If you're already building a custom version of TerraOS, ensure that swap is enabled by modifying the system's default kernel before building the image. The process for building a TerraOS image includes steps to chainload a regular Linux distro, which could be customized to include swap functionality by enabling it in the kernel and adding necessary utilities.

3. **Modify the RMA Shim**: Since both TerraOS and Shimboot exploit the Chromebook's RMA shims, another potential solution could involve unpacking and modifying the shim to allow swap. However, this approach is more complex and may face significant challenges due to the locked-down nature of ChromeOS environments. The custom shim would need to bypass kernel restrictions imposed by ChromeOS to re-enable features like swap.

### Conclusion:
To enable swap, your best option is to rebuild the kernel with swap enabled or create a custom bootloader and OS image that allows swap. However, keep in mind that working within the constraints of Chromebook recovery environments can be tricky due to system limitations designed for specific purposes like diagnostics and security【13†source】【12†source】.

**Reply by thatboyo** [reply]
@HelloIloveGimkit2 join relocate!!!

**Comment by FunnyCorgi** [comment]
for chatgpt(on my end)'s third suggestion: https://chromium.googlesource.com/chromiumos/platform/factory/+/HEAD/setup/RMA_SHIM.md
or just mount it like is done during modification on the build scripts for terraos and shimboot

**Reply by c0des1ayr** [reply]
Yeah, problem with that is that without Google's magic signature, the RMA shims don't work on enrolled Chromebooks. Perhaps using `kexec` would work, but that would require a custom kernel module since `kexec` is disabled in the shim kernels.
